sql/core/src/main/scala/org/apache/spark/sql/execution/joins/BroadcastHashJoinExec.scala

```diff
@@ -385,55 +408,101 @@ case class BroadcastHashJoinExec(
     val (matched, checkCondition, _) = getJoinCondition(ctx, input)
     val numOutput = metricTerm(ctx, "numOutputRows")
 
-    if (uniqueKeyCodePath) {
-      val found = ctx.freshName("found")
-      s"""
-         |boolean $found = false;
-         |// generate join key for stream side
-         |${keyEv.code}
-         |// Check if the key has nulls.
-         |if (!($anyNull)) {
-         |  // Check if the HashedRelation exists.
-         |  UnsafeRow $matched = (UnsafeRow)$relationTerm.getValue(${keyEv.value});
-         |  if ($matched != null) {
-         |    // Evaluate the condition.
-         |    $checkCondition {
-         |      $found = true;
-         |    }
-         |  }
-         |}
-         |if (!$found) {
-         |  $numOutput.add(1);
-         |  ${consume(ctx, input)}
-         |}
-       """.stripMargin
+    if (isNullAwareAntiJoin) {
+      require(leftKeys.length == 1, "leftKeys length should be 1")
+      require(rightKeys.length == 1, "rightKeys length should be 1")
+      require(right.output.length == 1, "not in subquery hash join optimize only single column.")
+      require(joinType == LeftAnti, "joinType must be LeftAnti.")
+      require(buildSide == BuildRight, "buildSide must be BuildRight.")
+      require(SQLConf.get.nullAwareAntiJoinOptimizeEnabled,
+        "nullAwareAntiJoinOptimizeEnabled must turn on for BroadcastNullAwareHashJoinExec.")
+      require(checkCondition == "", "not in subquery hash join optimize empty condition.")
+
+      if (broadcastRelation.value.inputEmpty) {
+        s"""
+           |// singleColumn NAAJ inputEmpty(true) accept all
+           |$numOutput.add(1);
+           |${consume(ctx, input)}
+         """.stripMargin
+      } else if (broadcastRelation.value.anyNullKeyExists) {
+        s"""
+           |// singleColumn NAAJ inputEmpty(false) anyNullKeyExists(true) reject all
+         """.stripMargin
+      } else {
+        val found = ctx.freshName("found")
+        s"""
+           |// singleColumn NAAJ inputEmpty(false) anyNullKeyExists(false)
+           |boolean $found = false;
+           |// generate join key for stream side
+           |${keyEv.code}
+           |// Check if the key has nulls.
+           |if (!($anyNull)) {
+           |  // Check if the HashedRelation exists.
+           |  UnsafeRow $matched = (UnsafeRow)$relationTerm.getValue(${keyEv.value});
+           |  if ($matched != null) {
+           |    $found = true;
+           |  }
+           |} else {
+           |  $found = true;
+           |}
+           |
+           |if (!$found) {
+           |  $numOutput.add(1);
+           |  ${consume(ctx, input)}
+           |}
+        """.stripMargin
+      }
     } else {
```

Comment:
nit: we can use `else if` to reduce code diff

Reply:
I am curious if it might be possible to apply the NAAJ special handling before, and then fall back onto the regular hash probing logic (that has an optimization for isUniqueKeyCodePath). Something like this:

(This only holds true for the single key case)

- Check NAAJ special cases: Build side has null, Build side is genuinely empty.
- If the stream side has a null key, ignore that stream row.
- Then probe the HT regularly and if there is any match, then don't return the row from the stream side. Else return.

My point is that the uniqueKeyCodePath pertains to the last step only of the steps above and perhaps does not need to be handled separately whether NAAJ is enabled or not.

I haven't thought through this, so it may not work at all :-P

Reply:
i will try do the refine.